docs/architecture.md

```diff
@@ -0,0 +1,176 @@
+
+2. The Graphsync requestor implementation makes requests to the network and handles incoming GraphSync responses. 
+
+3. The Graphsync responder implementation handles incoming GraphSync requests from the network and generates responses. 
+
+4. The message sending layer manages sending messages to peers. It is shared by both the requestor implementation and the responder implementation
+
+go-graphsync also depends on the following external dependencies:
+
+1. A network implementation, which provides basic functions for sending and receiving messages on a network.
+
+2. A bridge interface to `go-ipld-prime`, the library used to interact with IPLD data structures and perform selector queries. 
+
+3. A local blockstore implementation, expressed by a `loader` function and a `storer` function. 
+
+## Request Lifecycle
+
+In order to make a complete round trip GraphSync request all of the following need to happen:
+
+- Requestor needs to encode and send the request to the responder
+- Responder needs to receive the request, and perform an IPLD Selector query
+based on it
+- Responder needs to load blocks from local storage so that IPLD can perform the selector query
+- Responder needs to encode and send blocks traversed and metadata about the traversal to the requestor
+- Requestor needs to verify the blocks received are actually the right ones for the selector query requested. To do that it performs Selector Query locally, feeding it responses from the network.
+- Requestor needs to store blocks it receives once they are verified
+- Requestor needs to return traversed nodes to the Graphsync caller
+
+This order of these requirements corresponds roughly with the sequence they're executed in time.
+
+However, if you reverse the order of these requirements, it becomes clear that a GraphSync request is really an IPLD Selector Query performed locally that happens to be be backed by another remote peer performing the same query on its machine and feeding the results to the requestor.
+
+Selector queries, as implemented in the `go-ipld-prime` library, rely on a loader function to load data any time a link boundary is crossed during a query. The loader can be configured each time a selector query is performed. We use this to support network communication on both sides of a GraphSync query.
+
+On the requestor side, instead of supplying the local storage loader, we supply it with a different loader that waits for responses from the network -- and also simultaneously stores them in local storage as they are loaded. Blocks that come back on the network that are never loaded as part of the local Selector traversal are simply dropped. Moreover, we can take advantage of the fact that blocks get stored locally as they are traversed to limit network traffic -- there's no need to send back a block twice cause we can safely assume in a single query, once a block is traversed once, it's in the requestors local storage.
```

Comment:
```suggestion
On the requestor side, instead of supplying the local storage loader, we supply it with a different loader that waits for responses from the network -- and also simultaneously stores them in local storage as they are loaded. Blocks that come back on the network that are never loaded as part of the local Selector traversal are simply dropped. Moreover, we can take advantage of the fact that blocks get stored locally as they are traversed to limit network traffic -- there's no need to send back a block twice because we can safely assume in a single query, once a block is traversed once, it's in the requestors local storage.
```